### DISCUSSION TITLE
RFC: async defaultValues

### DISCUSSION THREAD

**Discussion by bluebill1049** [react-hook-form/react-hook-form#9046] [RFC]
## Context

The current status of setting `defaultValues` from async data source is using `reset`:

```ts
useEffect(() => {
  reset({
     firstName: 'test',
  })
}, [reset]);
```

honestly, this is probably not a pretty and clean approach... a few weeks before we had a feature request regarding async `defaultValues`: https://github.com/react-hook-form/react-hook-form/issues/2335

## Proposal Solutions

**Option A**

`defaultValues` accept `Promise`, and the promise will get resolved, although I do have concern over the confusion about which prop supports both `object` and `Promise`.

What's happening here: 

- promise will get resolved 
- `reset` API will get invoked to reset the form and hence provide reset options

```ts
const { formState: { isLoading } } = useForm({
  defaultValues: async () => {
    return {
      values: { firstName: 'test' },
      resetOptions: {
        keepDirtyValues: true,
      }
    }
  }
})
```

------------

**Option B**

- Reactive `values` for form values update
- Supplied with reset option to refine form values update behaviour

```tsx
const { data: values } = useQuery/useSWR()

useForm({ 
  values
  resetOptions: {
    keepDirtyValues: true,
  } 
})
```

happy to discuss more options on the table too.

Update 27/10/2022: going to update and support both `values` and async `defaultValues`

## POC

PR: https://github.com/react-hook-form/react-hook-form/pull/9052

cc @leapful @Moshyfawn

**Comment by JeromeDeLeon** [comment]
With `async defaultValues`, we are expecting it to do any asynchronous task right? So what if you rely on something else to trigger the `defaultValues` fetching? Like props? How would would I do that? Still I would need to rely to this `reset` + `useEffect` convention. Correct me if I'm wrong with all my understanding because that's how I see it. It really feels to me that I would still need to comeback to that approach.

**Comment by pmaier983** [comment]
I think this is a great idea. During my day job we work with large data sets, and we have specific queries to get available data for dropdown fields. If this was implemented it would pave the way to have a loading indicator and then some dropdown field all built without a Controller. 

The Async Promise seems like the best way forward (to me).

**Comment by bluebill1049** [comment]
Thanks @JeromeDeLeon I think you can pass arguments into that Promise callback.

```ts
useForm({
  defaultValues: async (args) => {
    if (args.test) { return { data: 'test' }; }
  }
})
```

happy for alternative solutions as well.

I think i see what do you mean now... yes defaultValue will be executed once at the first invoke.

**Comment by JeromeDeLeon** [comment]
Got it so its's not just about the API or anything. That makes sense at least to that part 😃

**Comment by adhrinae** [comment]
Do we need to provide another state like `isDefaultValueLoading`? Because for the first time users cannot see default values until that promise is resolved.

**Comment by bluebill1049** [comment]
> Do we need to provide another state like `isDefaultValueLoading`? Because for the first time users cannot see default values until that promise is resolved.

Good point probably will leave the user to define that state. `isLoading`.

`useState(false)`

**Comment by JeromeDeLeon** [comment]
You mean let the user define `useState` every time they will use `async defaulValues`? I think we should leverage that for users. just my opinion instead of letting the users define it every time

**Comment by keiya01** [comment]
```js
useForm({
  defaultValues: async (args) => {
    if (args.test) { return { data: 'test' }; }
  }
})
```

~~I think async function is invoked every render. By rendering, a lot of extra process is occurred.~~
This is wrong ...
Therefore I think we should provide other function that can invoke in `useEffect()`.

**Comment by JeromeDeLeon** [comment]
unless we wrap it with `useRef`, am I right @kotarella1110 ?

**Comment by kotarella1110** [comment]
> > Do we need to provide another state like `isDefaultValueLoading`? Because for the first time users cannot see default values until that promise is resolved.
> 
> Good point probably will leave the user to define that state. `isLoading`.
> 
> `useState(false)`

Is it possible to do the following?

```ts
useForm({
  defaultValues: async () => {
    const result = await fetch("xxxxx").then(res => res.json());
    return result;
  }
})
```

**Comment by JeromeDeLeon** [comment]
That's my question earlier, because what if you rely on something like `fetchById`?

**Reply by kellyrmilligan** [reply]
or like react-query

**Comment by bluebill1049** [comment]
You can still display your form, and wait for the promise to resolved and inject to all the inputs. if you need to show a loading spinner. you can just do a `useState`.

**Comment by bluebill1049** [comment]
```ts
useForm({
  defaultValues: async () => {
    updateLoading(false)
    const result = await fetch("xxxxx").then(res => res.json());
    updateLoading(true)
    return result;
  }
})
```
that works?

**Comment by bluebill1049** [comment]
I guess `callback` is just one solution, open to better options.

**Comment by keiya01** [comment]
Does this case work ?

```
useForm({
  defaultValues: {
    test: async (args) => 'test'
  }
})
```

**Comment by JeromeDeLeon** [comment]
```js
const [isLoading, setLoading] = useState(false);

useEffect(() => {
  async function asyncFetch() {
    setLoading(false);
    const result = await fetch(`xxxxx/${id}`).then(res => res.json());
    setLoading(true);
    reset(result);
  }

  asyncFetch();
}, [reset, id]);
```

Isn't it the same as the one above?

or we could just add `setDefaultValues`.

**Comment by bluebill1049** [comment]
Take a look at this issue https://github.com/react-hook-form/react-hook-form/issues/2333 @JeromeDeLeon I think you will have some context.

**Comment by JeromeDeLeon** [comment]
Sure. Thanks @bluebill1049 sorry for just popping out my opinion and all.

**Comment by bluebill1049** [comment]
not at all @JeromeDeLeon 👍

**Comment by bluebill1049** [comment]
> Does this case work ?
> 
> ```
> useForm({
>   defaultValues: {
>     test: async (args) => 'test'
>   }
> })
> ```

That works too :) I think would be good to solve a form level, individual input can be solved by `setValue`.

**Comment by keiya01** [comment]
I think we should provide `setDefaultValues` method because callback solution make it complex when it is bellow 🤔
`setValues` invoke `watch` method.

```js
useForm({
  defaultValues: {
    test: async (args) => {
      if (args.test) { 
        return { 
          data: 'test', 
          test:  {
            value: 'value'
          };
        } 
      }
      // ... some value or callback
    }
})
```

**Comment by bluebill1049** [comment]
or make `reset` API works with suspense: https://github.com/react-hook-form/react-hook-form/issues/2333

**Comment by JeromeDeLeon** [comment]
Is it not possible to work `setValue` and `reset` with `Suspense`?

**Comment by JeromeDeLeon** [comment]
> or make `reset` API works with suspense: #2333

:laughing: I didn't see this

**Comment by bluebill1049** [comment]
This: https://codesandbox.io/s/currying-cloud-d8e9f?file=/src/App.js

workaround version: https://codesandbox.io/s/interesting-waterfall-vg4lf?file=/src/App.js (think you will get the idea, react throw the Promise).

**Comment by smashercosmo** [comment]
I would just prefer form to always reinitialise when defaultValues change. Fetching default values asynchronously when initialising the form is less common scenario than just using props.

This is more common
```js
function EditUser(props) {
  const { user } = props
  const form = useForm({ defaultValues: user })
}
```

than
```js
function EditUser(props) {
  const { userId } = props
  const form = useForm({ defaultValues: async () =>  fetch(`/api/user/${userId}`)})
}
```

**Comment by bluebill1049** [comment]
Hey, @smashercosmo that's how I designed when hook form started, however this doesn't work well with 

```
function EditUser(props) {
  const { user } = props
  const form = useForm({ defaultValues: { pureObject: 'xxx' } })
}
```

unless user is willing to `memo` their `defaultValues` (which is not great DX), otherwise react dep check is always going to fail and each re-render will result in update value again.

**Comment by bluebill1049** [comment]
I think we can conclude `const form = useForm({ defaultValues: async () =>  fetch(`/api/user/${userId}`)})` is not great solution :(

**Comment by smashercosmo** [comment]
> otherwise react dep check is always going to fail and each re-render will result in update value again

And what's exactly bad about it?

**Comment by bluebill1049** [comment]
It's not Controller form, `defaultValues` is meant to set up the initial values. If every renders update input values, it defeats the purpose.

**Comment by smashercosmo** [comment]
Maybe we could introduce another property - "values"? so "defaultValues" will still be memoized and "values" will reinit the form every time it changes?

**Comment by bluebill1049** [comment]
Thanks @smashercosmo that's one potential solution as well.

**Comment by adhrinae** [comment]
> Maybe we could introduce another property - "values"? so "defaultValues" will still be memoized and "values" will reinit the form every time it changes? 

Sounds like his suggestion is dropping the advantage of uncontrolled component, but we can provide the option that our users can choose. 🙂

**Comment by bluebill1049** [comment]
I think we will keep this issue open for another week and gather all the input and feedback for the feature request. Probably will not be worked on right away.

**Comment by bluebill1049** [comment]
another option and i think it's a better one.

```ts
useForm({
  values: useMemo(() => values, [values]),
})
```

internally we will check values gets changes

```ts
useEffect(() => {
  reset(values)
}, [values])
```

**Comment by zulaica** [comment]
Just commenting to say I would _love_ something to be implemented here. I just ran into issues with this today. 😅

**Comment by dfyz011** [comment]
Such a pain to work with big forms in React, waiting for some good approach. Thank you

**Comment by bluebill1049** [comment]
> Such a pain to work with big forms in React, waiting for some good approach. Thank you

sorry, I think I miss read your message. I think what I purposed above is most likely going to be the final API, and you will have to memorize `values`.

we will do `reset` internally and depend on `value` reference update.

**Comment by jhonattansolartearanda** [comment]
This code in React Native, dont work.


`
   const {control, handleSubmit, errors} = useForm({
      defaultValues: async () => {
         return await prefsManager.getValue(enumerator.Url);
      },
   });
`

**Comment by bluebill1049** [comment]
> This code in React Native, dont work.
> 
> `const {control, handleSubmit, errors} = useForm({ defaultValues: async () => { return await prefsManager.getValue(enumerator.Url); }, });`

there is no such feature in the lib at the moment.

**Comment by vivek1996** [comment]
> another option and i think it's a better one.
> 
> ```ts
> useForm({
>   values: useMemo(() => values, [values]),
> })
> ```
> 
> internally we will check values gets changes
> 
> ```ts
> useEffect(() => {
>   reset(values)
> }, [values])
> ```


Hi, 
Thanks for the awesome library. 
So, this is the solution for setting the default values on async request?

**Comment by bluebill1049** [comment]
@vivek1996 right now yes, useEffect with reset

**Comment by beeirl** [comment]
@bluebill1049 I tried your suggested approach and called `reset()` in `useEffect` when I receive the async defaultValues. But this results in an infinite loop of `useEffect` for some reason. I've created a [minimalistic example](https://codesandbox.io/s/react-hook-form-get-started-forked-eoett?file=/src/index.js) to illustrate whats going on (in the example I don't use any type of async data, but the error stays the same). Please just comment in line 22 in the csb.

**Comment by guidalleprane** [comment]
> another option and i think it's a better one.
> 
> ```ts
> useForm({
>   values: useMemo(() => values, [values]),
> })
> ```
> 
> internally we will check values gets changes
> 
> ```ts
> useEffect(() => {
>   reset(values)
> }, [values])
> ```

As I understand it, then 'values' would be just one more prop to pass, right? This option seems to bring very little gain. Since we have the submission of the form being handled, why not the load of initial data?

I may be mixing things up, but for me it would make a lot of sense to get 'default values' asynchronously, it could have a `formState.isLoading` or something (just like with `form.handleSubmit` & `formState.isSubmitting`).

It would be a good way for the child components knows if the form data is still loading and display a skeleton component (especially if we are using `FormProvider`), in the same way that we already have `formState.isSubmitting` to disable a submit button for example.

Example: 

```ts
export default () => {
    const { formState, register } = useFormContext()

    return (
        {formState.isLoading 
             ? <Skeleton />
             : <input name="test" ref={register} />
        }
    )
}
```
Just with Suspense, I believe this would only be possible with a completely separate skeleton of entire page, which would not be very reusable. 

What do you think?

**Comment by vvo** [comment]
Small context, I came here because I am using https://swr.now.sh/ along with react-hook-form.
I am trying to wrap my mind around how to:
1. fetch data with swr (this I know how to do it)
2. solely use defaultValues with data coming from swr

The issue is that on first load the data will be "undefined" and then will become available. Example:

```js
  const { data } = useSWR("/api/user");

  const form = useForm({
    defaultValues: {
      name: user?.name
    },
  });
```

I hope this can help to find the right API for some sort of async defaultValues. Thanks!

**Comment by bluebill1049** [comment]
> Small context, I came here because I am using https://swr.now.sh/ along with react-hook-form.
> I am trying to wrap my mind around how to:
> 
> 1. fetch data with swr (this I know how to do it)
> 2. solely use defaultValues with data coming from swr
> 
> The issue is that on first load the data will be "undefined" and then will become available. Example:
> 
> ```js
>   const { data } = useSWR("/api/user");
> 
>   const form = useForm({
>     defaultValues: {
>       name: user?.name
>     },
>   });
> ```
> 
> I hope this can help to find the right API for some sort of async defaultValues. Thanks!

Thanks @vvo for the information. `defaultValues`is meant to set up `defaultValue` for each input, which is why it's cache the result, just like `<input defaultValue="xxx" />`. We will keep looking at this, probably not going to change `defaultValues`. behavior.

**Comment by vvo** [comment]
@bluebill1049 After posting the comment I realized that the best course of action, if I want to have defaultValues with the "right" data (= not undefined), is to wait to get the data to instantiate the form. This is definitely fine and I can show a skeleton while waiting for the data.

Thanks again for such a great lib

**Comment by vvo** [comment]
For anyone still coming here to learn how to have async defaultValues fields with react hook form, here's what I did following the author advice:

```jsx
function AsyncField() {
  const { register, reset, getValues } = useFormContext();
  const { asynchronousFieldValue } = useSomeAsynchronousDataHook();

  useEffect(() => {
    if (asynchronousFieldValue === undefined) {
      return; // loading
    }

    reset({
      ...getValues(),
      asynchronousFieldName: asynchronousFieldValue,
    });
  }, [reset, asynchronousFieldValue, getValues]);

  if (asynchronousFieldValue === undefined) return "Loading ...";

  return (
	<input
      type="text"
      defaultValue={asynchronousFieldValue}
      name="asynchronousFieldName"
      register={register}
      registerOptions={{ required: true }}
    />
  );
}
```

By using `...getValues()` you're able to reset all fields to the initial value while also adding your own asynchronous field value.

This way, isDirty will not be confused.

**Comment by bluebill1049** [comment]
Thank you @vvo very much for always sharing solutions here. ❤️

**Comment by antoinerousseau** [comment]
I'm using custom fields and a form provider. Had to add this to my Form for it to properly reset/populate when default values would change (data loading, or going from editing an object to creating a new one: same form but empty).

```ts
useEffect(() => {
  form.reset({ ...defaultValues } as DefaultValues<T>)
}, [defaultValues]) // eslint-disable-line react-hooks/exhaustive-deps
```

**Comment by dlqqq** [comment]
The proposed solution to this has a gotcha: if you update the default value object, you are resetting the entire form to the default values. Obviously, you don't want to do this while a user is potentially entering data. Thus, any async default values should be resolved prior to initial render of the form, with a `loading` catch block like:

```
if (query.loading) {
  return <Loading />
}

return <Form />
```

This kind of sucks since, intuitively, changing the default values of a form should not reset the entire form state. However, after reading through the RFC comments, I honestly can't think of a better solution either.

**Comment by zeddz92** [comment]
If you already set the defaultValue for each input like this
```
<input {...register("lastName")} defaultValue={user?.lastName}/>
```

You can just call `reset()` on useEffect and it will detect the changes
```
  useEffect(() => {
    if (user) {
      reset();
    }
  }, [user]);
```

**Comment by antoinerousseau** [comment]
So, in the end, we have no easy way to update pristine default values after the form was mounted?
I though `reset(defaultValues, { keepDirty: true})` was the way to go but [apparently not...](https://github.com/react-hook-form/react-hook-form/discussions/6415)

**Comment by MilkiasTonji** [comment]
> 
@zeddz92
Thank you very much!. This saved my day

**Comment by ScreamZ** [comment]
What's the best way to achieve this in 2022 ?

First I thought the best way is to not mount (and call useForm) as long data are not fetched.

**Comment by tiavina-mika** [comment]
I'm facing the same issue with `useForm`.

here is my code
```
...
 const currentUser = useSelector(getUserUserSelector);
...
 // LOAD DATA ON PAGE LOAD
 useEffect(() => {
    dispatch(getCurrentUser());
 }, [dispatch])
 
 ...
 const form = useForm<any>({
    defaultValues: {
      birthday: dayjs().subtract(18, 'year').toDate(),
      sex: 'Male',
      firstName: currentUser.firstName,
      lastName: currentUser.lastName,
    },
    resolver: zodResolver(signUpInfoSchema),
 });
```
 At first `currentUser `is `undefined`, then it will be loaded with async data,
 but the `useForm ``defaultValues `always get the first `undefined `data

**Comment by tiavina-mika** [comment]
here is my solution with useForm:

```
const loading = useSelector(getAppLoadingSelector);
const currentUser = useSelector(getUserUserSelector);

// LOAD DATA ON PAGE LOAD
useEffect(() => {
  dispatch(getCurrentUser());
}, [dispatch]);

const form = useForm<any>({
 // no need to define the default values here,
 // we define it in a useEffect (below)
  resolver: zodResolver(signUpInfoSchema),
});

// LOAD FORM DEFAULT VALUES
useEffect(() => {
  const defaultValues = getProfileInitialValues(currentUser);
  form.reset(defaultValues);
}, [currentUser, form]);

if (loading) return null;

```

```
/**
 * form initial values
 * @returns
 */
export const getProfileInitialValues = (user: User): SignupInfosInitialFormValues => {
  return {
    birthday: dayjs().subtract(18, 'year').toDate(),
    sex: SexEnum.MALE,
    firstName: user?.firstName,
    lastName: user?.lastName,
  };
};
```

**Comment by cbdeveloper** [comment]
Any updates on this?

I'm not a big fan of the `useEffect + reset` approach.

My current solution is to add a mid-level wrapper (to make sure async data is present).

And call the `useForm` hook one level down, where I'm sure async data is already present (because of the wrapper).

**Reply by abdullah-umer-dubizzle** [reply]
based

**Comment by mymactive** [comment]
React 18 releases a new function "Suspense".
And more, Next.js release [a new document](https://beta.nextjs.org/docs/getting-started) includes a new strategy for handling async data like [Data Fetching Fundamentals](https://beta.nextjs.org/docs/data-fetching/fundamentals).
I guess that a paradigm shift would occur for async data in React.
I also am interested in [a new React RFC "use"](https://github.com/reactjs/rfcs/pull/229).
We should be better think async defaultValues after async data handling is sold in React.

Sorry, no concrete idea.
I'm a big fan of RHF.
Thanks for developing.

**Reply by bluebill1049** [reply]
Update the proposal and PR: https://github.com/react-hook-form/react-hook-form/pull/9261 have a look at it and let us know your thoughts

**Comment by bluebill1049** [comment]
Beta: https://github.com/react-hook-form/react-hook-form/discussions/9381

**Comment by soonbenjy** [comment]
Hello !
Thanks for this new async defaultValues feature :) 

Did you consider handling error state during async fetch ?
`isLoading` boolean has been added inside `formState`, but nothing about a potential `isError` if fetch failed.

If you have an idea on how can we handle it, I would be pleased 🙏

**Reply by bluebill1049** [reply]
we can consider that for the latter, but it's probably easier to to try and catch that and use this feature when it's coming out: https://github.com/react-hook-form/react-hook-form/discussions/9691

**Comment by terragady** [comment]
How to handle errors during async function? When I do async and try to catch the error then type of defaultValues is screaming that `Type 'void' is not assignable to type 'FormSchema'.ts(2322)`
It works fine if I throw new Error but I want to handle error with errorLog function and then navigate to error page.

**Comment by yann-lauwers** [comment]
Thanks for this, it will be very helpful in the future! 

Just wondering why should I return an object that matches my schema when using async defaultValues ?
Returning an object without async doesn't throw an error even if my defaultValues object doesn't match the zod resolver.
Returning the same object with async on the default values throws and error.
![CleanShot 2024-01-12 at 12 01 11](https://github.com/react-hook-form/react-hook-form/assets/76515349/15542220-9229-40ac-adfc-526c99b3e9bd)
![CleanShot 2024-01-12 at 12 01 26](https://github.com/react-hook-form/react-hook-form/assets/76515349/e4d45857-2e00-4737-9467-f10bfdb5f2e1)

**Reply by yann-lauwers** [reply]
Is it normal the `type AsyncDefaultValues<TFieldValues>` is a `Promise<TFieldValues>;` instead of a `Promise<DefaultValues<TFieldValues>>;`.

As far as I can see in the types, default values either take `DefaultValues<TFieldValues>` either `AsyncDefaultValues<TFieldValues>`
[  defaultValues: DefaultValues<TFieldValues> | AsyncDefaultValues<TFieldValues>;](https://github.com/react-hook-form/react-hook-form/blob/7459afef82e5579ffa0486405a5ac448dec8b026/src/types/form.ts#L114C48-L114C66)

Thx for you help guys

**Reply by bryanprimus** [reply]
I was thinking the same; it should be `Promise<DefaultValues<TFieldValues>>` since `defaultValues` is optional from the start.

**Comment by gnunesdev** [comment]
If anyone came here wondering what would be the best option to handle async default values that came from a react-query hook for example, the recommendation is to use the `values` property, as described here: https://x.com/HookForm/status/1593066930579333120

**Comment by baruchiro** [comment]
There are two options suggested here:

1. use `async defaultValues` to fetch the data asyncronly.
2. use `values` to set reactive values.

There is no solution to set the default values for values that come from another hook (like `useQuery`).

**Reply by nick-carma** [reply]
it seems to me that if you set values, calling reset() later will use the most recent values.

**Comment by arvilmena** [comment]
is there a native way to handle errors due when the async defaultValue errored out? i.e, network timeout, server returned 400+ code?

```
  const { isLoading, errors } = form.formState;
```

i use `isLoading` to check if the defaultValues are being fetch, which is working well, but I cannot find a formState that points to when the defaultValues errored out. At the moment, I have to set a client boolean state for when it errored out.

**Reply by faustinozanetto** [reply]
Same issue here

**Comment by vvo** [comment]
@bluebill1049 Should we update the first post here to include this was released? https://github.com/react-hook-form/react-hook-form/pull/9261

**Comment by tomsotte** [comment]
Has the `Suspense` support been evaluated for async `defaultValues`? I think it would be great to have it

AFAIK it doesn't seem to be working; I've made a small [demo](https://stackblitz.com/edit/react-hook-form-suspense-defaultvalues?file=src%2FApp.tsx) here on how I think it might work, let me know if it's wrong or could be improved

**Reply by yann-lauwers** [reply]
It works fine, just need to add the loading as the form is loading.
Cf: https://stackblitz.com/edit/vitejs-vite-77d5kivv?file=src%2FApp.tsx

**Reply by tomsotte** [reply]
Isn't the loading state supposed to be managed implicitly by the Suspense? The use of form state `isLoading` it's like setting a flag inside the defaultValues when it's still loading

I've updated the demo with your code as well; as you can see the Suspense fallback is never rendered

**Reply by yann-lauwers** [reply]
Here is the doc to suspense:
```
Only Suspense-enabled data sources will activate the Suspense component. They include:

- Data fetching with Suspense-enabled frameworks like [Relay](https://relay.dev/docs/guided-tour/rendering/loading-states/) and [Next.js](https://nextjs.org/docs/app/building-your-application/routing/loading-ui-and-streaming#streaming-with-suspense)
- Lazy-loading component code with [lazy](https://react.dev/reference/react/lazy)
- Reading the value of a cached Promise with [use](https://react.dev/reference/react/use)
```
There is none of that in the useForm I believe.

**Reply by tomsotte** [reply]
Ok, thanks for your answer. What I wanted to ask was supporto for Suspense in useForm, in particular in async defaultValues.

**Reply by yann-lauwers** [reply]
You might maybe open a new ticket. That is a good idea!
I think this one isn't really observed anymore.

**Reply by tomsotte** [reply]
I've opened a feature request https://github.com/react-hook-form/react-hook-form/issues/13032 🚀
